Add white and blue panel variants

frankieroberto opened the pull request:
This adds 2 new variants of the Panel component, one with a solid blue background, and one with a white background but a blue border.

An additional `headingClasses` param is also added to enable a smaller title size class (`nhsuk-panel__title--s`).

## Blue panel

<img width="989" alt="Screenshot 2025-03-21 at 15 55 38" src="https://github.com/user-attachments/assets/d61b3604-4b9d-4c31-b882-36fceda6d5a5" />

This is intended for use when you need to stop users in a journey to tell them important information. This should be used vary sparingly, as there is some evidence that users can end up ignoring it, however it could work in some contexts, such as when shown after many regular pages with the off-white background.

The blue panel would be used with the [reversed Button variant](https://service-manual.nhs.uk/design-system/components/buttons#reverse-button) - which already mentions ‘interruption cards’ in the guidance.

See relevant threads for examples and discussion:

* [Stop users in a journey to tell them important information](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/241) on NHS design backlog
* [Interruption card](https://github.com/alphagov/govuk-design-system-backlog/issues/27) on GOVUK design backlog
* [Interruption card](https://design-patterns.service.justice.gov.uk/components/interruption-card/) in MoJ Design System

## White panel

<img width="1030" alt="Screenshot 2025-03-21 at 16 03 50" src="https://github.com/user-attachments/assets/2917424f-b8c0-49cb-ae56-69edd824d0db" />

This is intended for secondary content on the page, and would appear above the main content, such as an invitation to complete a survey, or to submit feedback about something you’ve just done.

It could be contain the upcoming new [Secondary button style](https://github.com/nhsuk/nhsuk-frontend/pull/1077).

See example: [Add feedback panel](https://github.com/NHSDigital/record-a-vaccination-prototype/pull/249):

<img width="601" alt="Screenshot 2025-03-21 at 16 10 50" src="https://github.com/user-attachments/assets/fff8eaae-ed8e-4cf5-806d-a69e6e6b817d" />

## Checklist

- [ ] Tested against our [testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [ ] Follows our [coding standards and style guide](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/coding-standards.md)
- [ ] CHANGELOG entry

frankieroberto commented on app/pages/examples.njk:
```diff
@@ -106,6 +106,8 @@
     <li><a href="../components/pagination/index.html">Pagination</a></li>
     <li><a href="../components/panel/index.html">Panel</a></li>
+    <li><a href="../components/panel/blue.html">Panel - blue</a></li>
+    <li><a href="../components/panel/white.html">Panel - white</a></li>
```
Done, ta.

frankieroberto commented on packages/components/panel/_panel.scss:
```diff
@@ -42,13 +42,28 @@ $nhsuk-border-width-panel: nhsuk-spacing(1);
   margin-top: 0;
 }
 
+.nhsuk-panel__title--s {
```
Yeah, maybe?!

I did think that maybe the existing `nhsuk-heading-s` class would work, but that gets overridden by `nhsuk-panel__title` (probably source order) so the panel titles need their own modifiers - which is probably better anyway?

frankieroberto replied:
not sure! `nhsuk-panel__title--s` _feels_ more better intuitive somehow, especially with title vs heading - but I could be persuaded either way.

anandamaryon1 replied:
Agree, if we're going to include heading sizes, it makes sense to include `s`, `m` and `l`.

frankieroberto replied:
I’ve dropped this for now, and removed the headingClasses param, as I'm not sure it’s needed for the Interruption card (default h1 size should be ok).

paulrobertlloyd commented:
Initial gut reaction; yes to the blue variant, however the white variant is something that appears to be _very_ different, and a component that possibly overlaps with a few others (card, notifications, warnings…).

frankieroberto commented:
@paulrobertlloyd the white one is definitely 'experimental'... 😂 

The [NHS warning callout](https://service-manual.nhs.uk/design-system/components/warning-callout) is quite different, and we don’t have a notification component (yet - maybe we should). The [Card component](https://service-manual.nhs.uk/design-system/components/card) is quite similar though (basically a different border) - but maybe that’s more designed for where the whole thing is clickable and you have several of them tiled up? (That said, there is a non-clickable one which I forgot about...)

frankieroberto commented:
Here’s our feedback panel using the existing card rather than the white panel variant:

<img width="968" alt="Screenshot 2025-03-21 at 17 03 45" src="https://github.com/user-attachments/assets/4163814d-17a6-4932-a081-5da389c50ade" />

Kinda works, but maybe stands out a bit less? 🤔

michaelgallagher commented:
re: the white card, worth considering [this](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/504)?

re: the blue card, minor thing, but should there be less padding between the button and bottom of the card?

frankieroberto commented:
> re: the white card, worth considering [this](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/504)?

Ooh I hadn’t seen that. Although in our case we’d need to use the secondary button as we already have a green primary button further down the page. And to me the blue border pairs nicely with the new blue outlined secondary button.

> re: the blue card, minor thing, but should there be less padding between the button and bottom of the card?

Yeah I think so. It’s caused by the extra default spacing below the button - but that can be reduced.

anandamaryon1 commented:
Nice, blue panel for interruption cards is helpful. With the minor tweak to button margin-bottom, `nhsuk-u-margin-bottom-3` seems about right to my eyes (accounting for the button shadow and the heading text increasing the padding at the top visually). 

Inclined to agree with others that the white variant doesn't seem to fit with the panel component.

sarawilcox commented:
I think the blue one is fine. We've had teams using something like this for interruption cards for some time. Like other people, I'm not sure about the white one.

frankieroberto commented:
@anandamaryon1 @paulrobertlloyd @colinrotherham @sarawilcox @michaelgallagher ok I’ve updated this pull request to purely add the blue panel variant (aka Interruption card).

Also reduced the margin below the button, although this is purely done using an `nhsuk-u-margin-bottom-3` override class rather than being automatic.

As the interruption card would normally be the only thing on the page, I think there’s no immediate need to support `headingClasses` in order to reduce the font size of the h1?

anandamaryon1 commented:
Looking at this and reading some backlog discussions about interruption cards, I've seen some suggestions to use start pages instead, as a way to break up a journey and provide some new context (e.g https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/241#issuecomment-632587931 & https://github.com/nhsuk/nhsapp-frontend/issues/19#issuecomment-2501391420). These seem reasonable suggestions. 

They would not be suitable for what I'd think of as 'system' messages, where the system has to tell you something, e.g something has changed and now you need to do something, but in those cases a notification banner seems more appropriate?

The difference between this interruption card and a 'normal' page is just the visual treatment, which makes you stop and think twice. I'm wondering whether we have any evidenced use-cases for this?

frankieroberto commented:
@anandamaryon1 I think there’s definitely a valid debate about whether you need the different visual treatment of a blue background, or whether just a regular content page with a green button with a distinct label is sufficient.

I don’t have any strong evidence for the need for the blue background personally, but others might? I think the MOJ Design system team did some research on it (see [their page](https://design-patterns.service.justice.gov.uk/components/interruption-card/)) and @edwardhorsford might remember the research from back on the passports service?

There would likely have to be some clear guidance to encourage people not to over-use it though, as I suspect part of the value of the blue background is it being distinct from the usual pattern and infrequently seen?

Do we know why the ["reverse" button](https://service-manual.nhs.uk/design-system/components/buttons#reverse-button) was first added to the NHS design system? The guidance for that already mentions interruption cards, but no other context offered!

> White buttons on solid background colour are good on components like interruption cards where link text, primary and secondary buttons would be lost.

paulmsmith commented:
111 online has the following patterns I would say fits the pattern use-case of "need to stop users in a journey to tell them important information."

**Module Zero**

<img width="1248" height="1280" alt="image" src="https://github.com/user-attachments/assets/a03d2282-c371-4c60-8e14-3a8075776573" />

**Some notes about this pattern:** 

- Referred to internally as Module zero or "Mod Zero", that name is taken from the Pathways software tool which is organised into 'modules'. 
- I think when describing it in our terms it is probably better called something like the 'vital to confirm' interrupt or even just as the content says the "Check it's not an emergency" interrupt.
- Research showed the need for a strong visual difference from the rest of the flow.
- It is actually a bit of a hack/work-around because of how long it would take to walk people though those conditions/causes-for-concern page-by-page.there was an assumption that people in a life threatening emergency were unlikely to go online
- Mod zero content is there as a "just in case"

I'm going to invite @LaurenMcAllister to contribute about her work on what we loving call the 'weird questions' interrupt.

colinrotherham commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
 
+## Unreleased
+
+### New features
```
I kept the emoji (and the bold is redundant, but a different colour in code)

```suggestion
### :new: **New features**
```

colinrotherham commented on CHANGELOG.md:
Do you mind undoing the "Trim Trailing Whitespace" changes?

Probably keep it turned off and let Prettier handle it?

We've used it in a couple of places to [preserve line breaks](https://www.markdownguide.org/basic-syntax/#line-break-best-practices)

paulrobertlloyd commented:
I’m for us adding this in, but I think the design is the easiest part of adding this, whereas getting the guidance right will be crucial. It’s odd that this is such a well used pattern across government, but only added to the MoJ Design System. Maybe their guidance can be a starting point for ours?

colinrotherham commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
+
+#### Blue panel variant
+
+There is a new variant of the Panel component with a solid blue background and white text. This can be used as an interruption card. [Pull request #1196](https://github.com/nhsuk/nhsuk-frontend/pull/1196)
```
I think we lowercase the component names when inline?

Looks good with the PR link split out for consistency

```suggestion
We've added a new variant of the panel component with a solid blue background and white text. This can be used as an interruption card.
  
This was added in [pull request #1196: Add blue panel (Interruption card) variant](https://github.com/nhsuk/nhsuk-frontend/pull/1196).
```

frankieroberto commented on packages/nhsuk-frontend/src/nhsuk/components/panel/macro-options.mjs:
```diff
@@ -76,6 +76,16 @@ export const examples = {
+  blue: {
+    context: {
+      titleText: 'You are doing something unusual',
+      html: '<p>Some text to explain.</p><a href="#" class="nhsuk-button nhsuk-button--reverse" data-module="nhsuk-button">Continue anyway</a>',
```
Done in 6d348781129e5464796c05308d0382bdb5405192

frankieroberto commented on packages/nhsuk-frontend/src/nhsuk/components/panel/_index.scss:
```diff
@@ -50,6 +50,10 @@ $nhsuk-border-width-panel: $nhsuk-panel-border-width;
     }
   }
 
+  .nhsuk-panel--blue {
```
I wasn't sure what to call it either. `--interruption` would get my vote I think.

It will probably get used for non-interruption purposes, but maybe in using the `--interruption` modifier and having strong guidance we can try and limit its use? (Given that part of the point is for it to be unusual)

frankieroberto commented:
I've just pushed b5ee59a8d03c50f1b4709b7cc54d90bf41808934 which adds support for a `titleClasses` macro param and a `nhsuk-panel__title--l` modifier which drops the font size down from 48 (XL) to 36 (L).

The XL size seems ok for the green confirmation panel, which typically uses a short h1 like "Application complete", but for the blue interruption panel the text could be quite a bit longer - eg see [MOJ examples](https://design-patterns.service.justice.gov.uk/components/interruption-card/).

colinrotherham commented:
Would be great to add as many examples to this PR as we can

frankieroberto commented:
@anandamaryon1 @colinrotherham @paulrobertlloyd think I’ve addressed all the comments!?

We can keep collecting examples in https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/241

I've added a 'Cancel' link beside the button in the example, but we can update the wording if we have a better example.

In terms of flexibility, I’ve added support for the `titleClasses` macro option, and `--xl`, `--l` and `--m` options for the title. Is that enough for now do we think?

colinrotherham reviewed (approved):
Just merged in `main` so we can see the new examples page:
https://nhsuk-frontend-br-add-panel-va.herokuapp.com/nhsuk-frontend/components/panel/

Looking good! Approved ✅

Before merging, did we get any real service examples?

E.g. Like the one from MoJ

<img width="1899" height="1163" alt="Image" src="https://github.com/user-attachments/assets/d84227c2-3ce8-4046-84ce-2966c6d36233" />

frankieroberto commented:
@sarawilcox most of those examples are from the app and are perhaps a slightly different use case?

We could do something vaccinations related?

> (patient name) had a COVID-19 vaccine less than 3 months ago
> 
> They had a COVID-19 vaccine on (date).
> 
> For most people, the minimum recommended gap between COVID-19 vaccine doses is 3 months.

sarawilcox commented:
> @sarawilcox most of those examples are from the app and are perhaps a slightly different use case?
> 
> We could do something vaccinations related?
> 
> > (patient name) had a COVID-19 vaccine less than 3 months ago
> > They had a COVID-19 vaccine on (date).
> > For most people, the minimum recommended gap between COVID-19 vaccine doses is 3 months.
> > [Continue anyway]

Yes, good to have a vaccine/staff facing one. I think this is a good example. The one above (Paul's) is a bit complex for a coded example.

frankieroberto commented:
@colinrotherham:

> Thank you for adding the example. Does the spacing below the link need another look?

That’s caused by the default margin below links in the button group. I've had a look at tweaking that separately in #1609

frankieroberto commented:
@anandamaryon1 @colinrotherham is this mergeable?

sarawilcox commented:
Hi Frankie, might be best if @colinrotherham @anandamaryon1 and I discuss this at our Monday PR catchup.

anandamaryon1 commented on packages/nhsuk-frontend/src/nhsuk/components/panel/template.njk:
```diff
@@ -5,7 +5,7 @@
   {%- if params.classes %} {{ params.classes }}{% endif %}"
   {{- nhsukAttributes(params.attributes) }}>
-  <h{{ headingLevel }} class="nhsuk-panel__title">
+  <h{{ headingLevel }} class="nhsuk-panel__title {%- if params.titleClasses %} {{ params.titleClasses }}{% endif %}">
```
Nested looks nicer to me. Could it / would it still support classes on the title?

frankieroberto replied:
Ok I've added `titleSize` for now. `titleClasses` will also work.

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/panel/template.njk:
Would you mind pulling out all the variables so they're set at the top?

We've not done it to all components yet, but you'll see this elsewhere:

```njk
{% from "nhsuk/macros/attributes.njk" import nhsukAttributes %}

{#- Set classes for this component #}
{%- set classNames = "nhsuk-panel" -%}
{%- set classNamesTitle = "nhsuk-panel__title" -%}

{%- if params.titleSize and
  params.titleSize in ["s", "m", "l", "xl"] and (
    not params.titleClasses or (
      not "nhsuk-panel__title--s" in params.titleClasses and
      not "nhsuk-panel__title--m" in params.titleClasses and
      not "nhsuk-panel__title--l" in params.titleClasses and
      not "nhsuk-panel__title--xl" in params.titleClasses
    )
  )
%}
  {% set classNamesTitle = classNamesTitle ~ " nhsuk-panel__title--" ~ params.titleSize %}
{% endif %}

{%- if params.classes %}
  {% set classNames = classNames ~ " " ~ params.classes %}
{% endif %}

{%- if params.titleClasses %}
  {% set classNamesTitle = classNamesTitle ~ " " ~ params.titleClasses %}
{% endif %}

{%- set headingLevel = params.headingLevel if params.headingLevel else 1 -%}
```

colinrotherham commented on packages/nhsuk-frontend/src/nhsuk/components/panel/template.njk:
```diff
@@ -1,11 +1,33 @@
-  {%- if params.classes %} {{ params.classes }}{% endif %}"
-  {{- nhsukAttributes(params.attributes) }}>
-  <h{{ headingLevel }} class="nhsuk-panel__title">
+{# Set classes for this component #}
```
```suggestion
{#- Set classes for this component #}
```

colinrotherham replied:
Resolved in https://github.com/nhsuk/nhsuk-frontend/pull/1196/commits/3a5dabb60f4d6cd83df5020f309a5a999ef35c77

colinrotherham commented:
I've brought the examples over from https://github.com/nhsuk/nhsuk-service-manual/pull/2316

* [**Panel component** example listing page](https://nhsuk-frontend-br-add-panel-va.herokuapp.com/nhsuk-frontend/components/panel/)

Can update them as necessary
